File: sample/objective-c/GNAdSampleMultipleNative/GNAdSampleMultipleNative/TableViewController.m
//
//  TableViewController.m
//  GNAdSampleMultipleNative


#import "TableViewController.h"
#import "TableViewCell.h"
#import "MyCellData.h"
#import "GNQueue.h"

@interface TableViewController ()
{
    BOOL _loading;
    NSTimeInterval secondsStart, secondsEnd;
    GNQueue *queueAds; // into AdObject.
}

@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *indicator;
@property (strong, nonatomic) NSMutableArray *cellDataList;
@end

static const int CELLCOUNT = 20;

@implementation TableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    
    secondsStart = [NSDate timeIntervalSinceReferenceDate];
    _cellDataList = [NSMutableArray array];
    queueAds = [[GNQueue alloc] initWithMaxSize:50];
    
    // Create GNNativeAdRequest.
    nativeAdRequest = [[GNNativeAdRequest sharedMenager] initWithID:@"YOUR_SSP_APP_ID"];
    nativeAdRequest.delegate = self;
    nativeAdRequest.GNAdlogPriority = GNLogPriorityInfo; // show log.
    nativeAdRequest.geoLocationEnable = YES;
    
    // ad loading.
    [nativeAdRequest multiLoadAds];
}

- (void)dealloc
{
    nativeAdRequest.delegate = nil;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark - GNNativeAdRequestDelegate

// Called when GNNativeAdRequest native ads request succeeded.
- (void)nativeAdRequestDidReceiveAds:(NSArray *)nativeAds
{
    secondsEnd = [NSDate timeIntervalSinceReferenceDate];
    NSLog(@"TableViewController: nativeAdRequestDidReceiveAds in %f seconds.",(double)(secondsEnd - secondsStart));
    for (GNNativeAd *nativeAd in nativeAds) {
        [queueAds enqueue:nativeAd];
    }
    [self createCellDataList];
    
}

// Called when GNNativeAdRequest native ads request failed.
- (void)nativeAdRequest:(GNNativeAdRequest *)request didFailToReceiveAdsWithError:(NSError *)error
{
    NSLog(@"TableViewController: didFailToReceiveAdsWithError : %@", [error localizedDescription]);
    
    [self createCellDataList];
}

// Sent before ad begins open landingURL in External Browser.
- (BOOL)shouldStartExternalBrowserWithClick:(GNNativeAd *)nativeAd landingURL:(NSString *)landingURL
{
    NSLog(@"TableViewController %s : %@",__FUNCTION__,landingURL);
    return YES;
}

#pragma mark - Get My Cell Data

- (void)requestCellDataListAsync
{
    _loading = YES;
}

- (void)createCellDataList
{
    for (int i = 0; i < CELLCOUNT; i++) {
        if ([queueAds count] > 0) {
            // ad object.
            id ad = [queueAds dequeue];
            [_cellDataList addObject:ad];
        } else {
            // No ad object.
            [_cellDataList addObject:[[MyCellData alloc] init]];
        }
    }
    [_indicator stopAnimating];
    [self.tableView reloadData];
    _loading = NO;
    
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return _cellDataList.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"SampleDataCell";
    TableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    
    cell.iconView.layer.borderColor = [UIColor blackColor].CGColor;
    cell.iconView.layer.borderWidth = 1;
    cell.iconView.layer.cornerRadius = 10;
    cell.iconView.layer.masksToBounds = YES;
    
    if ([[_cellDataList objectAtIndex:indexPath.row] isKindOfClass:[GNNativeAd class]]) {
        // GNNativeAd Cell.
        GNNativeAd *nativeAd = (GNNativeAd *)[_cellDataList objectAtIndex:indexPath.row];
        if (nativeAd != nil) {
            cell.nativeAd = nativeAd;
            cell.titleLabel.text = nativeAd.title;
            cell.descriptionLabel.text = nativeAd.description;
            cell.iconView.image = nil;
            NSURL *url = [NSURL URLWithString:nativeAd.icon_url];
            // NSURL *url = [NSURL URLWithString:nativeAd.screenshots_url];
            [TableViewController requestImageWithURL:url completion:^(UIImage *image, NSError *error){
                if (error) return;
                
                [cell.iconView setImage:image];
            }];
            // tracking point.
            [nativeAd trackingImpressionWithView:cell];
        }
    } else {
        // Not GNNativeAd Cell
        MyCellData *mycellData = (MyCellData *)[_cellDataList objectAtIndex:indexPath.row];
        cell.nativeAd = nil;
        cell.titleLabel.text = [mycellData.title stringByAppendingFormat:@" No.%ld",indexPath.row + 1];
        cell.descriptionLabel.text = mycellData.content;
        cell.iconView.image = nil;
        NSURL *url = mycellData.imgURL;
        
        [TableViewController requestImageWithURL:url completion:^(UIImage *image, NSError *error){
            if (error || ![url isEqual:mycellData.imgURL]) return;
            cell.iconView.image = image;
        }];
    }
    return cell;
}

#pragma mark - Scroll view delegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    CGFloat contentOffsetWidthWindow = self.tableView.contentOffset.y + self.tableView.bounds.size.height;
    BOOL leachToBottom = contentOffsetWidthWindow >= self.tableView.contentSize.height;
    if (!leachToBottom || _loading) return;
    [_indicator startAnimating];
    secondsStart = [NSDate timeIntervalSinceReferenceDate];
    [nativeAdRequest multiLoadAds]; // load ad.
    [self requestCellDataListAsync];
}

#pragma mark - Table View Click

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    TableViewCell *cell = (TableViewCell*)[tableView cellForRowAtIndexPath:indexPath];
    if (cell.nativeAd != nil) {
        [cell.nativeAd trackingClick:cell];
    } else {
        [self performSegueWithIdentifier:@"selectRow" sender:self];
    }
}

#pragma mark - Request and Creqte Icon Image

+ (void)requestImageWithURL:(NSURL*)url completion:(void (^)(UIImage *image, NSError *error))completion
{
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    NSURLSessionConfiguration *sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:sessionConfiguration];
    
    [[session dataTaskWithRequest:request
                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error){
                    if (error) {
                        completion(nil,error);
                        return;
                    }
                    __block UIImage *image = [UIImage imageWithData:data];
                    dispatch_async(dispatch_get_main_queue(), ^{
                        completion(image,nil);
                        return;
                    });
                }] resume];
}
@end

File: sample/swift/GNAdSampleMultipleNative/GNAdSampleMultipleNative/TableViewController.swift
//
//  TableViewController.swift
//  GNAdSampleMultipleNative

import UIKit
import Foundation

class TableViewController: UITableViewController {
    @IBOutlet weak var indicator: UIActivityIndicatorView!
    var loading: Bool = false
    var secondsStart: TimeInterval = 0.0, secondsEnd: TimeInterval = 0.0
    var queueAds: GNQueue = GNQueue(aMaxSize: 50)
    var cellDataList: NSMutableArray = NSMutableArray()
    
    var nativeAdRequest: GNNativeAdRequest = GNNativeAdRequest(id: "YOUR_APP_SSP_ID")
    
    override func viewDidLoad() {
        super.viewDidLoad()
        secondsStart = Date.timeIntervalSinceReferenceDate
        
        // Load GNNativeAdRequest
        nativeAdRequest.delegate = self
        nativeAdRequest.gnAdlogPriority = GNLogPriorityInfo
        nativeAdRequest.geoLocationEnable = true
        loading = true

        nativeAdRequest.multiLoadAds()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Get My Cell Data
    
    func createCellDataList() {
        for _ in 0 ..< 10 {
            if queueAds.count() > 0 {
                let ad: AnyObject? = queueAds.dequeue()
                if ad != nil {
                    cellDataList.add(ad!)
                }
            } else {
                cellDataList.add(MyCellData())
            }
        }
        indicator.stopAnimating()
        self.tableView.reloadData()
        loading = false
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        // #warning Incomplete implementation, return the number of sections
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of rows
        return cellDataList.count
    }

    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "SampleDataCell", for: indexPath) as! TableViewCell
        let isAdCell = cellDataList.object(at: indexPath.row) is GNNativeAd
        
        cell.icon.layer.borderColor = UIColor.black.cgColor
        cell.icon.layer.borderWidth = 1
        cell.icon.layer.cornerRadius = 10
        cell.icon.layer.masksToBounds = true
        
        if isAdCell {
            let nativeAd: GNNativeAd = cellDataList.object(at: indexPath.row) as! GNNativeAd
            cell.nativeAd = nativeAd
            cell.title.text = nativeAd.title
            cell.content.text = nativeAd.description
            cell.icon.image = nil
            let url: URL = URL(string: nativeAd.icon_url)!
            requestImageWithURL(url: url, completion: {(image: UIImage!, error: NSError?) -> Void in
                if error != nil {
                    return
                }
                cell.icon.image = image
            })
            nativeAd.trackingImpression(with: cell)
        } else {
            let myCellData: MyCellData = cellDataList.object(at: indexPath.row) as! MyCellData
            cell.nativeAd = nil
            cell.title.text = (myCellData.title as String) + " No.\(indexPath.row + 1)"
            cell.content.text = myCellData.content as String
            cell.icon.image = nil
            let url: URL = myCellData.imgURL as URL
            requestImageWithURL(url: url, completion: {(image: UIImage!, error: NSError?) -> Void in
                if error != nil {
                    return
                }
                cell.icon.image = image
            })
        }
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let cell = tableView.cellForRow(at: indexPath) as! TableViewCell
        if cell.nativeAd != nil {
            cell.nativeAd.trackingClick(cell)
        } else {
            self.performSegue(withIdentifier: "selectRow", sender: self)
        }
    }
    
    // MARK: - UIScrollViewDelegate
    
    override func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let contentOffsetWidthWindow: CGFloat = self.tableView.contentOffset.y + self.tableView.bounds.size.height
        let heightContent: CGFloat = self.tableView.contentSize.height
        var leachToBottom: Bool = false
        if contentOffsetWidthWindow >= heightContent { leachToBottom = true }
        if (!leachToBottom || loading) { return }
        indicator.startAnimating()
        
        secondsStart = Date.timeIntervalSinceReferenceDate
        loading = true
        nativeAdRequest.multiLoadAds()
    }
    
    
    func requestImageWithURL(url: URL, completion:@escaping (_ image: UIImage?, _ error: NSError?)->Void) {
        
        let task = URLSession.shared.dataTask(with: url) { data, response, error in
            if let data = data, let response = response {
                print(response)
                let image: UIImage? = UIImage(data: data)
                DispatchQueue.main.async {
                    completion(image, error as NSError?)
                    return
                }
            } else {
                print(error ?? "Error")
            }
        }
        task.resume()
    }
}


// MARK: - GNNativeAdRequestDelegate

extension TableViewController: GNNativeAdRequestDelegate {
    
    func nativeAdRequestDidReceiveAds(_ nativeAds: [Any]!) {
        secondsEnd = Date.timeIntervalSinceReferenceDate
        NSLog("TableViewController: nativeAdRequestDidReceiveAds in %f seconds.", (Double)(secondsEnd - secondsStart));
        for nativeAd in nativeAds as! [GNNativeAd] {
            queueAds.enqueue(anObject: nativeAd)
        }
        createCellDataList()
    }
    
    func nativeAdRequest(_ request: GNNativeAdRequest!, didFailToReceiveAdsWithError error: Error!)
    {
        NSLog("TableViewController: didFailToReceiveAdsWithError : %@.", error.localizedDescription);
        createCellDataList()
    }
    
    func shouldStartExternalBrowser(withClick nativeAd: GNNativeAd!, landingURL: String!) -> Bool {
        NSLog("TableViewController: shouldStartExternalBrowserWithClick : %@.", landingURL);
        return true;
    }
}

